File: cpp/opencv/yolo_opencv_classify.cpp
/* 
 * @Author: taifyang
 * @Date: 2024-06-12 09:26:41
 * @LastEditTime: 2026-01-19 23:37:42
 * @Description: source file for YOLO opencv classification
 */

#include "yolo_opencv.h"

void YOLO_OpenCV_Classify::init(const Algo_Type algo_type, const Device_Type device_type, const Model_Type model_type, const std::string model_path)
{
	if (algo_type != YOLOv5 && algo_type != YOLOv8 && algo_type != YOLOv11 && algo_type != YOLOv12 && algo_type != YOLO26)
	{
		std::cerr << "unsupported algo type!" << std::endl;
		std::exit(-1);
	}
	YOLO_OpenCV::init(algo_type, device_type, model_type, model_path);
	YOLO_Classify::init(algo_type, device_type, model_type, model_path);
}

void YOLO_OpenCV_Classify::pre_process()
{
	cv::Mat crop_image;

	if (m_algo_type == YOLOv5)
	{
#ifdef OPENCV_WITH_CUDA
		cv::cuda::GpuMat gpu_image, gpu_crop_image, gpu_cvt_image;
    	gpu_image.upload(m_image);
		int crop_size = std::min(m_image.cols, m_image.rows);
		int left = (m_image.cols - crop_size) / 2, top = (m_image.rows - crop_size) / 2;
		gpu_crop_image = gpu_image(cv::Rect(left, top, crop_size, crop_size));
		cv::cuda::resize(gpu_crop_image, gpu_crop_image, cv::Size(m_input_size.width, m_input_size.height));
		gpu_crop_image.convertTo(gpu_cvt_image, CV_32FC3, 1. / 255.);
		cv::cuda::subtract(gpu_cvt_image, cv::Scalar(0.406, 0.456, 0.485), gpu_cvt_image);
		cv::cuda::divide(gpu_cvt_image, cv::Scalar(0.225, 0.224, 0.229), gpu_cvt_image);
		gpu_cvt_image.download(crop_image);
#else
		CenterCrop(m_image, crop_image);
		Normalize(crop_image, crop_image, m_algo_type);
#endif
	}
	else if (m_algo_type == YOLOv8 || m_algo_type == YOLOv11 || m_algo_type == YOLOv12 || m_algo_type == YOLO26)
	{	
#ifdef OPENCV_WITH_CUDA	
		cv::cuda::GpuMat gpu_image, gpu_crop_image, gpu_cvt_image;
		gpu_image.upload(m_image);
		if (m_image.cols > m_image.rows)
			cv::cuda::resize(gpu_image, gpu_image, cv::Size(m_input_size.height * m_image.cols / m_image.rows, m_input_size.height));
		else
			cv::cuda::resize(gpu_image, gpu_image, cv::Size(m_input_size.width, m_input_size.width * m_image.rows / m_image.cols));
		int crop_size = std::min(crop_image.cols, crop_image.rows);
		int left = (crop_image.cols - crop_size) / 2, top = (crop_image.rows - crop_size) / 2;
		gpu_crop_image = gpu_image(cv::Rect(left, top, crop_size, crop_size));
		cv::cuda::resize(gpu_crop_image, gpu_crop_image, cv::Size(m_input_size.width, m_input_size.height));
		gpu_crop_image.convertTo(gpu_cvt_image, CV_32FC3, 1. / 255.);
		gpu_cvt_image.download(crop_image);
#else
		if (m_image.cols > m_image.rows)
			cv::resize(m_image, crop_image, cv::Size(m_input_size.height * m_image.cols / m_image.rows, m_input_size.height));
		else
			cv::resize(m_image, crop_image, cv::Size(m_input_size.width, m_input_size.width * m_image.rows / m_image.cols));
		CenterCrop(m_image, crop_image);
		Normalize(crop_image, crop_image, m_algo_type);
#endif
	}

	cv::dnn::blobFromImage(crop_image, m_input, 1, cv::Size(crop_image.cols, crop_image.rows), cv::Scalar(), true, false);
}

void YOLO_OpenCV_Classify::post_process()
{
	m_output0_host = (float*)m_output[0].data;

	std::vector<float> scores(m_class_num); 
	float sum = 0.0f;
	for (size_t i = 0; i < m_class_num; i++)
	{
		scores[i] = m_output0_host[i];
		sum += exp(m_output0_host[i]);
	}
	int id = std::distance(scores.begin(), std::max_element(scores.begin(), scores.end()));

	m_output_cls.id = id;
	if (m_algo_type == YOLOv5)
		m_output_cls.score = exp(scores[id]) / sum;
	else if (m_algo_type == YOLOv8 || m_algo_type == YOLOv11 || m_algo_type == YOLOv12 || m_algo_type == YOLO26)
		m_output_cls.score = scores[id];

	if(m_draw_result)
		draw_result(m_output_cls);
}

File: python/backends/OpenCV/yolo_opencv_classify.py
'''
Author: taifyang
Date: 2024-06-12 22:23:07
LastEditTime: 2026-09-11 00:14:57
Description: opencv inference class for the YOLO classifaction algorithm
'''


from backends.utils import *
from backends.yolo_classify import *
from backends.OpenCV.yolo_opencv import *


'''
description: opencv inference class for the YOLO classifaction algorithm
'''     
class YOLO_OpenCV_Classify(YOLO_OpenCV, YOLO_Classify):
    '''
    description:            construction method
    param {*} self          instance of class
    param {str} algo_type   algorithm type
    param {str} device_type device type
    param {str} model_type  model type
    param {str} model_path  model path
    return {*}
    '''      
    def __init__(self, algo_type:str, device_type:str, model_type:str, model_path:str) -> None:
        YOLO_OpenCV.__init__(self, algo_type, device_type, model_type, model_path)
        YOLO_Classify.__init__(self, algo_type, device_type, model_type, model_path)

    '''
    description:    model pre-process
    param {*} self  instance of class
    return {*}
    '''    
    def pre_process(self) -> None:
        if self.algo_type in ['YOLOv5']:
            input = centercrop(self.image, self.inputs_shape)
            input = normalize(input, self.algo_type)
        elif self.algo_type in ['YOLOv8', 'YOLOv11', 'YOLOv12', 'YOLO26']:
            self.inputs_shape = (224, 224)
            if self.image.shape[1] > self.image.shape[0]:
                input = cv2.resize(self.image, (self.inputs_shape[0]*self.image.shape[1]//self.image.shape[0], self.inputs_shape[0]))
            else:
                input = cv2.resize(self.image, (self.inputs_shape[1], self.inputs_shape[1]*self.image.shape[0]//self.image.shape[1]))
            input = centercrop(input, self.inputs_shape)
            input = normalize(input, self.algo_type)
            
        input = input[:, :, ::-1].transpose(2, 0, 1)  #BGR2RGB and HWC2CHW
        self.inputs = np.expand_dims(input, axis=0).astype(dtype=np.float32)
        self.net.setInput(self.inputs)
    
    '''
    description:    model post-process
    param {*} self  instance of class
    return {*}
    '''    
    def post_process(self) -> None:
        output = np.squeeze(self.outputs).astype(dtype=np.float32)
        if self.algo_type in ['YOLOv5']:
            class_id = np.argmax(output)
            scores =np.exp(np.max(output))/np.sum(np.exp(output))
        elif self.algo_type in ['YOLOv8', 'YOLOv11', 'YOLOv12', 'YOLO26']:
            class_id = np.argmax(output)
            scores = np.max(output)

        if self.render_result:
            self.result = self.draw_result(class_id, scores)
